src/aiida_quantumespresso/workflows/xspectra/base.py

```diff
@@ -0,0 +1,263 @@
+        :param abs_atom_marker: the name given to the Kind representing the absorbing atom.
+                                Matches to Kind.name for the absorbing atom in the structure,
+                                used to set the parameter `xiabs` automatically.
+        :param protocol: protocol to use, if not specified, the default will be used.
+        :param overrides: optional dictionary of inputs to override the defaults of the protocol.
+        :param options: A dictionary of options that will be recursively set for the ``metadata.options`` input of all
+            the ``CalcJobs`` that are nested in this work chain.
+        :return: a process builder instance with all inputs defined ready for launch.
+        """
+        from aiida_quantumespresso.workflows.protocols.utils import recursive_merge
+
+        if isinstance(code, str):
+            code = orm.load_code(code)
+
+        type_check(code, orm.Code)
+
+        inputs = cls.get_protocol_inputs(protocol, overrides)
+
+        metadata = inputs['xspectra']['metadata']
+        parameters = inputs['xspectra']['parameters']
+
+        if options:
+            metadata['options'] = recursive_merge(inputs['xspectra']['metadata']['options'], options)
+
+        parent_calc = parent_folder.creator
+        if parent_calc.process_type == 'aiida.calculations:quantumespresso.xspectra':
+            inputs['xspectra']['kpoints'] = parent_calc.inputs.kpoints
+            xiabs = parent_calc.inputs.parameters.get_dict()['INPUT_XSPECTRA'].get('xiabs', 1)
+            parameters['INPUT_XSPECTRA']['xiabs'] = xiabs
+        elif parent_calc.process_type == 'aiida.calculations:quantumespresso.pw':
+            structure = parent_folder.creator.inputs.structure
+            kinds_present = sorted([Kind.name for Kind in structure.kinds])
+            if abs_atom_marker not in kinds_present:
+                raise AttributeError(
+                    f'given abs_atom_marker ({abs_atom_marker}) does not match a Kind in the structure.'
+                )
+            else:
+                parameters['INPUT_XSPECTRA']['xiabs'] = kinds_present.index(abs_atom_marker) + 1
+        else:
+            raise AttributeError(f'Parent calculation type ({parent_calc.process_type}) is not  suitable.')
+
+        # pylint: disable=no-member
+        builder = cls.get_builder()
+        builder.xspectra['code'] = code
+        builder.xspectra['parent_folder'] = parent_folder
+        builder.xspectra['core_wfc_data'] = core_wfc_data
+        builder.xspectra['parameters'] = orm.Dict(inputs['xspectra']['parameters'])
+        builder.xspectra['metadata'] = metadata
+        if 'settings' in inputs['xspectra']:
+            builder.xspectra['settings'] = orm.Dict(inputs['xspectra']['settings'])
+        builder.clean_workdir = orm.Bool(inputs['clean_workdir'])
+        if 'kpoints' in inputs['xspectra']:
+            builder.kpoints = parent_calc.inputs.kpoints
```

Comment:
Is this logic correct? Should it not be
```python
builder.kpoints = inputs['xspectra']['kpoints']
```
otherwise you are ignoring `inputs['xspectra']['kpoints']`

Reply:
Yes, this actually has been superseded by the changes made previously, but I forgot to remove it. As the check of the `parent_calc` now sets the logic for the `kpoints` input (use parent mesh if `XspectraCalculation` or use structure to generate a new mesh if `PwCalculation`), I think it would make sense to simply merge these two parts together:

        if parent_calc.process_type == 'aiida.calculations:quantumespresso.xspectra':
            inputs['xspectra']['kpoints'] = parent_calc.inputs.kpoints
            xiabs = parent_calc.inputs.parameters.get_dict()['INPUT_XSPECTRA'].get('xiabs', 1)
            parameters['INPUT_XSPECTRA']['xiabs'] = xiabs
        elif parent_calc.process_type == 'aiida.calculations:quantumespresso.pw':
            structure = parent_folder.creator.inputs.structure
            kinds_present = sorted([Kind.name for Kind in structure.kinds])
            if 'kpoints' in inputs['xspectra']:
                builder.kpoints = parent_calc.inputs.kpoints
            else:
                builder.kpoints_distance = orm.Float(inputs['kpoints_distance'])
                builder.kpoints_force_parity = orm.Bool(inputs['kpoints_force_parity'])

Reply:
If it makes sense to have a different mesh for the `xspectra.x` code then is used for `pw.x` that makes sense to me.

Reply:
It is the case that some users might still want to (for instance) use a denser mesh for the `xspectra.x` calculation than the `pw.x` calculation - particularly some of the published works and tutorial examples that I've seen do this as well - so allowing this does indeed make sense.